src/rust/engine/logging/src/logger.rs

```diff
@@ -185,51 +203,43 @@ impl Log for PantsLogger {
     let destination = stdio::get_destination();
 
     // Build the message string.
-    let log_string = {
-      let mut log_string = {
-        let cur_date = chrono::Local::now();
-        format!(
-          "{}.{:02}",
-          cur_date.format(TIME_FORMAT_STR),
-          cur_date.time().nanosecond() / 10_000_000 // Two decimal places of precision.
-        )
-      };
-
-      let use_color = destination.stderr_use_color();
-
-      let level = record.level();
-      let level_marker = match level {
-        _ if !use_color => format!("[{}]", level).normal().clear(),
-        Level::Info => format!("[{}]", level).normal(),
-        Level::Error => format!("[{}]", level).red(),
-        Level::Warn => format!("[{}]", level).yellow(),
-        Level::Debug => format!("[{}]", level).green(),
-        Level::Trace => format!("[{}]", level).magenta(),
-      };
-      write!(log_string, " {}", level_marker).unwrap();
-
-      if inner.show_target {
-        write!(log_string, " ({})", record.target()).unwrap();
-      };
-      writeln!(log_string, " {}", log_msg).unwrap();
-      log_string
+    let log_time = {
+      let cur_date = chrono::Local::now();
+      format!(
+        "{}.{:02}",
+        cur_date.format(TIME_FORMAT_STR),
+        cur_date.time().nanosecond() / 10_000_000 // Two decimal places of precision.
+      )
+    };
+    let log_line = if inner.show_target {
+      format!("{} ({})", log_msg, record.target())
+    } else {
+      log_msg
     };
-    let log_bytes = log_string.as_bytes();
+    let uncolored_log_string = fmt_log!(log_time, self.get_level_marker(false, record), log_line);
+    let uncolored_log_bytes = uncolored_log_string.as_bytes();
 
     {
-      let mut maybe_per_run_file = inner.per_run_logs.lock();
+      let mut maybe_per_run_file = inner.per_run_log.lock();
       if let Some(ref mut file) = *maybe_per_run_file {
         // deliberately ignore errors writing to per-run log file
-        let _ = file.write_all(log_bytes);
+        let _ = file.write_all(uncolored_log_bytes);
       }
     }
 
+    let stderr_log_string = if destination.stderr_use_color() {
+      fmt_log!(log_time, self.get_level_marker(true, record), log_line)
+    } else {
+      uncolored_log_string.clone()
+    };
```

Comment:
Probably a micro-optimization. Alternatively we can just do `fmt_log!(log_time, self.get_level_marker(destination.stderr_use_color(), record), log_line)`.

Dealer's choice.